### PR TITLE
feat: added ability to capture both fields and args seen during an Apollo query as metrics

### LINKED ISSUE #202: Capture metrics for every field being requested in a graphql query
## Is your feature request related to a problem? Please describe.
As you maintain a graphql service you'll likely refactor and want to remove older fields.  The Alerts team wants the ability to know every field requested per request so they know if a field is no longer being asked for so they can safely remove.

## Feature Description
At the end of every request report metrics for every unique field being requested within a graphql query.

## Additional context
We currently report scalars as resolve segments but this occurs for every single item in a request.  

## Priority
Please help us better understand this feature request by choosing a priority from the following options:
[Nice to Have, Really Want, Must Have, Blocker]

### PR DESCRIPTION
**Pull request by bizob2828**
## Proposed Release Notes
 * Added ability to capture both fields and args seen during an Apollo query as metrics.
   * To enable this feature, update your createPlugin configuration to specify `captureFieldMetrics: true`

## Links
Closes #202 

## Details
These metrics purely track call count unlike the `GraphQL/resolve` and `GraphQL/typedResolve` which track how long it takes for a resolver to execute.  I did this because if a customer just wants to track fields/args requested they wouldn't get them all as scalars are gated behind `captureScalars`. The way we capture scalars is to record a segment and I didn't want to incur this overhead for just tracking fields/args requested.  

Here's a query you can use to query these new metrics, which I added to the docs/metrics.md

```
FROM Metric SELECT count(newrelic.timeslice.value) where appName = '[YOUR APP NAME]' WITH METRIC_FORMAT 'GraphQL/{kind}/ApolloServer/{field}' where kind = 'arg' or kind = 'field' FACET kind, field limit max since 1 day ago 
```

### CODE REVIEW

File: lib/create-plugin.js [hunk 1]
```diff
@@ -175,9 +171,9 @@ function createPlugin(api, config = {}) {
               // the attributes to resolver segments as they help
               // inform performance impacts.
               for (const segment of [operationSegment, resolverSegment]) {
-                for (const [key, value] of Object.entries(args)) {
+                for (const [key, value] of Object.entries(flattenedArgs)) {
```
Comment (bizob2828) [inline lib/create-plugin.js:174]: this was actually a bug and now I realized I forgot to add a test demonstrating this

Comment (jordigh) [reply lib/create-plugin.js:174]: I'm interested in seeing that bug, please add the test.

File: lib/create-plugin.js [hunk 2]
```diff
@@ -570,4 +602,60 @@ function updateOperationSegmentName(context, operationSegment) {
...
+ * @param {Object} params.args args key from resolver context
+ * @param {Object} params.fieldType parent type of field
+ * @param {Object} params.fieldName name of field
+ *
+ */
+function captureFieldMetrics({ transaction, args, fieldType, fieldName }) {
+  const fieldMetric = `${FIELD_PREFIX}/${fieldType}.${fieldName}`
+  createCallCountMetric(transaction, fieldMetric)
+  Object.entries(args).forEach(([key]) => {
+    if (fieldName && fieldType) {
```
Comment (jordigh) [inline lib/create-plugin.js:654]: What is this check for? What would we record for `fieldMetric` above if either of these is falsey?

Comment (bizob2828) [reply lib/create-plugin.js:654]: shit, this was applicable before i moved things around

File: tests/versioned/attributes-tests.js [hunk 3]
```diff
@@ -366,6 +366,42 @@ function createAttributesTests(t) {
...
+      t.match(resolveAttributes, expectedArgAttributes)
+      t.match(operationAttributes, expectedArgAttributes)
+    })
+
+    executeQuery(serverUrl, query, (err) => {
+      t.error(err)
+      t.end()
+    })
+  })
+
```
Comment (jordigh) [inline tests/versioned/attributes-tests.js:404]: Okay, I get it, before we'd be trying to add an object as an attribute value, which isn't valid. So we just weren't adding any nested args at all.

**Review by jordigh — CHANGES_REQUESTED** [review]

File: tests/versioned/attributes-tests.js (hunk 3 above)
Comment (bizob2828) [reply tests/versioned/attributes-tests.js:404]: exactly

File: lib/create-plugin.js [hunk 4]
```diff
@@ -304,6 +300,35 @@ function flattenToArray(fieldPath) {
...
+ * @param {array} [seen=[]] objects that have been seen
+ */
+function flattenArgs({ result = {}, prefix = '', obj, seen = [] }) {
+  seen.push(obj)
+
+  for (const key in obj) {
+    if (seen.indexOf(obj[key]) > -1) {
+      continue
+    }
+
```
[the remark this answers is not in the document]
Comment (bizob2828) [reply lib/create-plugin.js:321]: hmm, I copied this function from the agent. let me dig

[the remark this answers is not in the document]
Comment (bizob2828) [reply lib/create-plugin.js:321]: good catch, i removed this logic

**Pull request merged**